sky/resources.py

```diff
@@ -932,6 +940,29 @@ def _try_validate_ports(self) -> None:
         # We don't need to check the ports format since we already done it
         # in resources_utils.simplify_ports
 
+    def _try_validate_labels(self) -> None:
+        """Try to validate the labels attribute.
+
+        Raises:
+            ValueError: if the attribute is invalid.
+        """
+        if self._labels is None:
+            return
+
+        if self.cloud is None:
+            # Because each cloud has its own label format, we cannot validate
+            # the labels without knowing the cloud.
+            with ux_utils.print_exception_no_traceback():
+                raise ValueError(
+                    'Cloud must be specified when labels are provided.')
```

Comment:
Supporting failover is slightly tricky. 

For example, labels in Kubernetes commonly use `/` to create a "namespace" for tags (e.g., `skypilot.co/accelerators`, `app.kubernetes.io/component`.. see [recommended labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/common-labels/#labels)). However, GCP does not support `/` in labels, and a failover would cause provisioning to fail. Putting a stricter global format check would prevent users from creating these labels at all. 

Another option could be to not do any validation at all and let these checks fail at provisioning time. We could do that, but this seemed cleaner so went with this for now. Any other ideas? Happy to change if you think we should support failover with labels.

Reply:
I see! I think it should be fine to require the `cloud` to be presented when the `labels` is specified. Another option to support failover could be the following:

Validate the labels in the following function, and return an empty list if the labels do not match the requirement for a specific cloud, i.e. make the current cloud infeasible:
https://github.com/skypilot-org/skypilot/blob/431f567e5a6d97e057607762b05a9813990f6b2a/sky/clouds/cloud.py#L323-L356

Reply:
Agreed, that could work. Seems a little hacky though, since the error now becomes hidden in the warning instead of being a clear error.

Checking in resources.py (current):
```
sky launch task.yaml
Task from YAML spec: task.yaml
ValueError: Invalid label my/label=my/value. Invalid label value my/value for GCP. Value can include lowercase alphanumeric characters, dashes, and underscores, with a total length of 63 characters or less.
```

Checking in get_feasible_launchable_resources with warning (last proposal):
```
$ sky launch task.yaml
Task from YAML spec: task.yaml
W 04-30 15:36:34 cloud.py:383] Label my/label=my/value is invalid for cloud GCP. Reason: Invalid label value my/value for GCP. Value can include lowercase alphanumeric characters, dashes, and underscores, with a total length of 63 characters or less.
I 04-30 15:36:34 optimizer.py:1209] No resource satisfying GCP({'T4': 1}) on GCP.
sky.exceptions.ResourcesUnavailableError: Catalog does not contain any instances satisfying the request:
Task(run=<empty>)
  resources: GCP({'T4': 1}).

To fix: relax or change the resource requirements.

Hint: sky show-gpus to list available accelerators.
      sky check to check the enabled clouds.
```

wdyt? I'm leaning towards keeping the current variant, but can change to checking in `get_feasible_launchable_resources` to support failover if you feel strongly about it.

Reply:
Keeping the current way sounds good to me, but we may want to file an issue for it for supporting failover when the  labels are specified.

Reply:
Filed https://github.com/skypilot-org/skypilot/issues/3500!